packages/smooth_app/lib/pages/product/simple_input_page.dart

```diff
@@ -139,6 +140,24 @@ class _SimpleInputPageState extends State<SimpleInputPage> {
     }
   }
 
+  /// Merge the changes from [newProduct] into [productInCache].
+  Product _mergeProducts(Product productInCache, Product newProduct) {
```

Comment:
I'm not a big fan of that `_mergeProducts` method, to say the least.

With those "simple input" classes, we have a very powerful method called `getChangedProduct`, which "transforms" the input product according to the user changes.

A clean way of coding this PR would be:
* I land on a "simple" page.
* I edit some fields.
* I save the page
  * I want the product to be instantly locally refreshed
  * At the same time I queue a call to the server that changes the impacted fields
* For that I need to call `getChangedProduct` twice
  * First time from the product I fetch from the local database (get-product / change-product / save-product), and that I put in the `UpToDateProductProvider`
  * Second time from an empty product (just the barcode) (create-empty-product-with-barcode, change-product, queue-product-server-update)
* Et voilà!

If we agree on that, I suggest that you refactor the other input pages in order to include an `AbstractSimpleInputPageHelper`, in order to have the same process for all pages (basically, we always need a helper method that changes a product).

Reply:
Hello @monsieurtanuki 

I guess i get you what i wanted but before i commit  the changes i wanted to make sure this what you wanted 


```dart
    final Product changedProduct = Product(barcode: widget.product.barcode);
    final LocalDatabase localDatabase = context.read<LocalDatabase>();
    final AppLocalizations appLocalizations = AppLocalizations.of(context);
    final UpToDateProductProvider provider =
        context.read<UpToDateProductProvider>();
    final DaoProduct daoProduct = DaoProduct(localDatabase);
    final Product? cachedProduct = await daoProduct.get(
      changedProduct.barcode!,
    );
    bool changed = false;
    bool added = false;
    String pageName = '';
    for (int i = 0; i < widget.helpers.length; i++) {
      if (widget.helpers[i].addItemsFromController(_controllers[i])) {
        added = true;
      }
      if (widget.helpers[i].getChangedProduct(changedProduct)) {
        changed = true;
        if (cachedProduct != null) {
          widget.helpers[i].getChangedProduct(cachedProduct);
        }
      }
      pageName = widget.helpers[i].getTitle(appLocalizations);
    }
```
Full code here at https://pastebin.com/M5vqa12z

Reply:
It's more or less what I meant, except that if you write your code directly in a commit it's MUCH easier to me to understand what you wrote: here I have to parse the whole thing instead of only the differences.
So please commit your code.
An alternate way would be to exchange in an issue about the best possible solutions BEFORE even creating a PR. Less trouble for the developer and the reviewer.

Reply:
The code is now pushed, please review again